### ISSUE TITLE
Minimum dependencies to removed packages and minimum pre-release packages.

### ISSUE THREAD

**Issue by joelverhagen** [NuGet/Home#5882]
@Thorium commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663)

I don't know which Nuget-repository is correct for this bug, but at least this one has the issues tab.

I own a package called [Owin.Comperession](https://www.nuget.org/packages/Owin.Compression/). It has a following dependency:
`Owin (>= 1.0.0-prerelease && <= 1.0.0)`

And everything worked fine and everyone was happy. Until one day, the Owin-team unlisted the prerelease-version of the package from NuGet. Now my users started to get build errors.

1.0 is still there. Do you consider all pre-release versions are better than release versions?

A repository to reproduce the issue is here:
https://github.com/jomonson/OwinCompressionDependencyIssue

---

@joelverhagen commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-328885891)

Thanks for reporting this issue @Thorium. The available versions on nuget.org are 0.5.0, 0.7.0, 0.11.0, 0.12.0, 0.14.0, and 1.0.0. The 1.0.0 version of Owin is the only version that satisfies the version range in your Owin.Compression package. I am not familiar with a 1.0.0-prerelease version of the Owin package. Perhaps it was on some CI or secondary source like MyGet?

Concerning the build failure, I was only able to reproduce a build warning in Visual Studio 2017:
```
warning NU1603: Owin.Compression 1.0.16 depends on Owin (>= 1.0.0-prerelease && <= 1.0.0) but Owin 1.0.0-prerelease was not found. An approximate best match of Owin 1.0.0 was resolved.
```

This warning is just saying that 1.0.0-prerelease was not found so 1.0.0 was used instead. NuGet prefers the lowest version in the dependency version range. Is it possible you have warnings as errors enabled?

As a side note, you can test version range compatibility using this tool:
http://nugettoolsdev.azurewebsites.net/4.3.0-preview4/version-satisfies?versionRange=%5B1.0.0-prerelease%2C1.0.0%5D&version=1.0.0

---

@Thorium commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-329005118)

My customer says Owin.Compression 1.0.16 gives warning but 1.0.15 error, and it's the same dependency.

I have not used CI or MyGet or any other sources than NuGet.org

Why the warning? So should I release a new package for every removed minimum dependency, saying "no changes" in the release notes?

---

@joelverhagen commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-329010617)

> My customer says Owin.Compression 1.0.16 gives warning but 1.0.15 error, and it's the same dependency.

I am not sure why this is happening and have been unable to reproduce the error. Do you have steps?

> So should I release a new package for every removed minimum dependency, saying "no changes" in the release notes?

Yes, that's quite reasonable. 

---

@Thorium commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-329018310)

>> So should I release a new package for every removed minimum dependency, saying "no changes" in the release notes?

> Yes, that's quite reasonable

Ok... How do package owners get notifications of minimum dependencies not being available anymore?

---

@joelverhagen commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-329019074)

We do not have a notification mechanism for package deletion since package deletions are a very rare case. Our package deletion policy is here:
https://docs.microsoft.com/en-us/nuget/policies/deleting-packages

Given that 1.0.0 was published in 2012, I don't have any records of whether 1.0.0-prerelease ever existed at all on NuGet.org. 

---

@forki commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-329049163)

1.0.0 prerelease probably never existed but that's not the question here.

The range is >= 1.0.0-prerelease && <= 1.0.0 and there is a 1.0 of owin. 1.0 is clearly higher than any prerelease and definitely smaller or equal to 1.0. So this version clearly matches this range. Why is it even showing a warning (or crashes)? 

---

@joelverhagen commented on [Tue Sep 12 2017](https://github.com/NuGet/NuGetGallery/issues/4663#issuecomment-329049674)

All great questions! Let's move this issue over to the NuGet client repository (https://github.com/NuGet/Home/issues) so that the client guys can give all the details. Thanks for all the helpful information!

**Comment by zhili1208** [comment]
@mishra14 @emgarten

**Comment by emgarten** [comment]
Dupe of https://github.com/NuGet/Home/issues/5764

Please see the above issue for a full description of what NU1603 means.

**tl;dr:** NU1603 as a warning means that your packages could change if a lower version that matches the min version of the range shows up later in a source.

**Comment by forki** [comment]
Please reopen. There where errors reported

**Comment by emgarten** [comment]
@forki @Thorium would you point me to the error or some repro steps?

I tried using https://github.com/jomonson/OwinCompressionDependencyIssue with both 1.0.16 and 1.0.15, on my machine using the latest NuGet I only get an NU1603 warning. 

Most of the discussion here is around the NU1603 warning and how dependency ranges work, I think this is covered in https://github.com/NuGet/Home/issues/5764

The only mention of an error I see is here: 
> My customer says Owin.Compression 1.0.16 gives warning but 1.0.15 error, and it's the same dependency.

Trying out a packages.config project instead to see if there is a difference I do not see any issues. Owin 1.0.0 is installed since it is allowed by the dependency range. 

```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="FSharp.Core" version="4.1.17" targetFramework="net461" />
  <package id="Microsoft.Owin" version="3.1.0" targetFramework="net461" />
  <package id="Owin" version="1.0" targetFramework="net461" />
  <package id="Owin.Compression" version="1.0.15" targetFramework="net461" />
  <package id="System.Collections" version="4.0.11" targetFramework="net461" />
  <package id="System.Console" version="4.0.0" targetFramework="net461" />
  <package id="System.Diagnostics.Debug" version="4.0.11" targetFramework="net461" />
  <package id="System.Diagnostics.Tools" version="4.0.1" targetFramework="net461" />
  <package id="System.Globalization" version="4.0.11" targetFramework="net461" />
  <package id="System.IO" version="4.1.0" targetFramework="net461" />
  <package id="System.Linq" version="4.1.0" targetFramework="net461" />
  <package id="System.Linq.Expressions" version="4.1.0" targetFramework="net461" />
  <package id="System.Linq.Queryable" version="4.0.1" targetFramework="net461" />
  <package id="System.Net.Requests" version="4.0.11" targetFramework="net461" />
  <package id="System.Reflection" version="4.1.0" targetFramework="net461" />
  <package id="System.Reflection.Extensions" version="4.0.1" targetFramework="net461" />
  <package id="System.Resources.ResourceManager" version="4.0.1" targetFramework="net461" />
  <package id="System.Runtime" version="4.1.0" targetFramework="net461" />
  <package id="System.Runtime.Extensions" version="4.1.0" targetFramework="net461" />
  <package id="System.Runtime.Numerics" version="4.0.1" targetFramework="net461" />
  <package id="System.Text.RegularExpressions" version="4.1.0" targetFramework="net461" />
  <package id="System.Threading" version="4.0.11" targetFramework="net461" />
  <package id="System.Threading.Tasks" version="4.0.11" targetFramework="net461" />
  <package id="System.Threading.Tasks.Parallel" version="4.0.1" targetFramework="net461" />
  <package id="System.Threading.Thread" version="4.0.0" targetFramework="net461" />
  <package id="System.Threading.ThreadPool" version="4.0.10" targetFramework="net461" />
  <package id="System.Threading.Timer" version="4.0.1" targetFramework="net461" />
</packages>
```

I'm happy to investigate further and re-open this issue if someone could provide me with more details. As it looks right now NuGet client handles *Owin.Compression* and *Owin* correctly.
